### PR TITLE
Fix #76 (server event layer)

### LINKED ISSUE #76: server event layer
we need a programmable layer between the commands reception in routes, and the commands execution in model.

this layer encapsulates the calls to model methods. this will allow us to trigger custom pre and post actions. the new layer will propose methods to register and unregister event observers.

```
routes -> event layer -> model
```

we will decompose what will happen when receiving a command (eg model:create):

* the server receives a command which involves a model:create (create, upload or write)
* instead of calling model:create directly, we call the event dispatcher.
* the dispatcher has a stack of pre-command to trigger. triggers are done, each return value is analyzed. if a return value is false, the command is aborted.
* if all pre-command triggered are successful the dispatcher executes the command
* if successful, the registered post-command are triggered

```
command
route:create |
             | -> el:create -> model:create
route:upload |
```

this way we will be able to plug actions like permission checking, actions on the file system, socket.io message sending, mail, log, and pipeline-specific triggers.

Implementation note: the event observers stack should be treated with extreme care regarding the execution of custom actions and the return values, using a clear and simple async or sync scheme. The actions may be ran asynchronously using a nested scheme or asynchronously all ran at the same time. the developer will have to consider this with care to choose the best (simple, scalable) way.

### PR DESCRIPTION
**Pull request by gnoxr**
## UPDATE
See the next comments in order to stay up to date. The major part of this one has been stripped in order to keep the messages clean (I only kept the "Go further" and "Sample code" sections, to quickly show the differences between the two versions).

---

### Go further

- The comments are improvable with documentation-generators' syntax (`@something`).
- We could imagine a "priority" system, allowing to attach listeners while specifying their priority, to sort them and run them in a specific order.
- It should be better to `this.attach()` new listeners to the next run instead of the current one.
- If there are 3 times the same listener, and the last one calls `this.detach()`, it will detach the first one.
- It might be interesting to add support for regexes (ie, listen for `.*create.*` will listen simultaneously for `pre-create`, `created`, `create-comment`, etc).
- `interrupt()` is stupid when listeners run asynchronous code. We have to wait until the listener finished its processing - in other words, let each listener call the `next()` one.

### Sample code

```javascript
var events = require('./events');
var count = 0;

function testFire() {
    this.data[++count] = 'fired event ' + this.name;
}

function testArguments() {
    this.data[++count] = arguments;
}

function testDetach() {
    this.data[++count] = 'detaching from ' + this.name;
    this.detach();
}

function testAttach() {
    this.data[++count] = 'attaching testArguments to ' + this.name;
    this.attach(testArguments);
}

// Attach listeners

events.attach('test', testFire);
events.attach('test', testDetach);
events.attach('test', testAttach);
events.attach('test', testFire);

// Run tests

count = 0;
console.log(events.fire('test', 'arg10', 'arg11', 'arg12'));
console.log();

count = 0;
events.detach('test', testFire); // remove the first testFire listener
events.detach('test', testAttach); // remove the first testAttach listener
console.log(events.fire('test', 'arg20', 'arg21', 'arg22'));
console.log();
```
Output
```javascript
{ '1': 'fired event test',
  '2': 'detaching from test',
  '3': 'attaching testArguments to test',
  '4': 'fired event test',
  '5': { '0': 'arg10', '1': 'arg11', '2': 'arg12' } }

{ '1': 'fired event test',
  '2': { '0': 'arg20', '1': 'arg21', '2': 'arg22' } }
```
The listener '2' detached itself during the first run.
The listeners '1' and '3' from the first run are detached before the second run.

### CODE REVIEW

**Comment by gnoxr** [comment]
I updated the pull request description with the last changes and an example. I will now work on other issues while waiting for feedback.

**Comment by gnoxr** [comment]
These commits implement an "event layer". To use it, you need to read the documentation below and to instantiate an event manager using the following code:
```js
var events = require('./events.js');
```

Below are some details on this implementation.

### EventManager

The EventManager is an object managing events (obviously) and event listeners. It is the object actually returned when requiring `./events.js`. All files requiring it in a same Node instance will share the same events and listeners. It is actually the most simple event manager I could imagine, so there is room for adding more features (see the "Go further" section below).

It provides the following methods:

```javascript
EventManager.attach(hook, callback)
```
Attach the function `callback` as a listener of the named event `hook`. It will be called whenever the `hook` event is fired.

```javascript
EventManager.detach(hook, listener)
```
Remove the first occurrence of the `listener` from the named `hook`.

```javascript
EventManager.fire(hook[, ...])
```
Fire the `hook`, by calling asynchronously all of its listeners with the given optional arguments.

### Event context

A context (`ctx`) object is used as `this` for all listeners whenever they are called. The listeners can still access themselves using `this.listener`.

**Important** `this.data` is shared between all listeners in the current run. It will be returned by `fire` to the firer after running all listeners.

```javascript
this.detach()
```
Remove the current listener (without halting it) from the current hook. If there are multiple times the same listener, remove the first occurrence.

```javascript
this.attach(callback)
```
Add a new listener to the current hook. I do not remember testing it, but the newly attached callback should be called in the current firing among with the other listeners currently being run.

```javascript
this.next()
```
Call the next listener.

```javascript
this.end()
```
Put an end to the current run. No more listeners will be called afterwards (even when calling `this.next()`), therefore `fire` will return immediately to its `then` callback.

**Important** Read the example to understand the use of `then()`.

### Whys

There are some technical choices, here are the whys.

- In the initial commit, listeners were named (in an object). Why not now?

> To prevent possible conflicts. Listener names are some kind of pain to developers.

- Why didn't you use an already existing Node module?

> It was easier and faster to write and test this code than to download modules, read their documentation, test them, compare them together and choose the best one suiting our needs.

- Why do you loop through `arguments` using a numeric index instead of `Object.keys()`, as `arguments` is an object?

> The indexes of `arguments` are numeric, and using a numeric index is two orders of magnitude faster (I benchmarked using 10 arguments) than using `Object.keys(arguments)` or any `for(a of arguments)` or `for(a in arguments)` iterator, as they rely on `Object.keys`. Numeric indexes are efficient and adapted to the current problem; we can't pass an `arguments` object directly to the `apply` method.

- Why do you declare the context object inside of the `fire` method instead of keeping a separated class?

> Listeners should not have access to the `next` value. Having the object inside of this function allows some JavaScript scope hacks. This way we can keep `next` private while modifying it inside of the context methods. The code is also more concise this way, and we have only one class in the file.

- Will you try to prevent listeners from firing events?

> No. This kind of solution kills performance and we don't need this complexity, even to prevent infinite loop of auto-firing events. These problems are to be avoided during the development process of a listener.

- Why did you remove `interrupt()` to use `next()` instead?

> `interrupt()` is problematic when listeners run asynchronous code. We need to wait until the listener finished its processing before continuing running the code. In other words, each listener must indicate when we can call the `next()` one.
> Note that I put instead a `end()` method, which does the same thing.

### Go further

- The comments are improvable with documentation-generators' syntax (`@something`).
- We could imagine a "priority" system, allowing to attach listeners while specifying their priority, to sort them and run them in a specific order.
- It should be better to `this.attach()` new listeners to the next run instead of the current one.
- It might be interesting to add support for regexes (ie, listen for `.*create.*` will listen simultaneously for `pre-create`, `created`, `create-comment`, etc).

### Sample code

```javascript
var events = require('./events');
var count = 0;

function testFire() {
    this.data[++count] = 'fired event ' + this.name;
    this.next();
}

function testArguments() {
    this.data[++count] = arguments;
    this.next();
}

function testDetach() {
    this.data[++count] = 'detaching from ' + this.name;
    this.detach();
    this.next();
}

function testAttach() {
    this.data[++count] = 'attaching testArguments to ' + this.name;
    this.attach(testArguments);
    this.next();
}

// Attach listeners

events.attach('test', testFire);
events.attach('test', testDetach);
events.attach('test', testAttach);
events.attach('test', testFire);

// Run tests

count = 0;
events.fire('test', 'arg10', 'arg11', 'arg12').then(function (data) {
    console.log(data);
    console.log();

    count = 0;
    events.detach('test', testFire); // remove the first testFire listener
    events.detach('test', testAttach); // remove the first testAttach listener
    events.fire('test', 'arg20', 'arg21', 'arg22').then(function (data) {
        console.log(data);
        console.log();
    });
});
```
Output
```javascript
{ '1': 'fired event test',
  '2': 'detaching from test',
  '3': 'attaching testArguments to test',
  '4': 'fired event test',
  '5': { '0': 'arg10', '1': 'arg11', '2': 'arg12' } }

{ '1': 'fired event test',
  '2': { '0': 'arg20', '1': 'arg21', '2': 'arg22' } }
```
The listener '2' detached itself during the first run.
The listeners '1' and '3' from the first run are detached before the second run.

**Comment by remyla** [comment]
thank you very much.

the next step would be to implement a use case for the event system like the default keys on creation
these keys are created in every routes where a node creation is involved (/api/create, /api/version etc)
this could be interesting to move this default keys creation to an event listener

```js
       // Control properties
        var author = req.user.username || req.connection.remoteAddress;
        var time = Date.now();
        for (var n in nodes) {
            if ('object' !== typeof nodes[n]) {
                res.status(400);
                res.send('create error: the specified elements must be objects');
                return;
            }
            nodes[n].author = author;
            nodes[n].time = time;
        }
```

is this something feasible and consistent?
thank you

**Comment by gnoxr** [comment]
I will be working on it on a new branch.

**Comment by gnoxr** [comment]
About the database interface, we could use something like this:
```javascript
for (var f in db) {
    if (typeof db[f] === 'function') {
        DatabaseInterface[f] = function () {
            var callback = arguments[arguments.length - 1];
            arguments[arguments.length - 1] = function (err, doc) {
                events.fire(err ? f + '-fail' : f, doc);
                callback(err, doc);
            }
            db[f].apply(arguments);
        }
    }
}
```
In order to expose every public method of the DB.
